onnxruntime/core/providers/cuda/tensor/scatter_nd.cc

```diff
@@ -85,5 +101,88 @@ Status ScatterND::ComputeInternal(OpKernelContext* context) const {
   return Status::OK();
 }
 
+Status ScatterNDWithAtomicReduction::ComputeInternal(OpKernelContext* context) const {
+  const auto* input_tensor = context->Input<Tensor>(0);
+  const auto* indices_tensor = context->Input<Tensor>(1);
+  const auto* updates_tensor = context->Input<Tensor>(2);
+
+  const auto& input_shape = input_tensor->Shape();
+  const auto& indices_shape = indices_tensor->Shape();
+  const auto& updates_shape = updates_tensor->Shape();
+
+  // Validate input shapes
+  ORT_RETURN_IF_ERROR(onnxruntime::ScatterND::ValidateShapes(input_shape, indices_shape, updates_shape));
+
+  auto* output_tensor = context->Output(0, input_shape);
+
+  const void* input_data = input_tensor->DataRaw();
+  void* output_data = output_tensor->MutableDataRaw();
+
+  if (input_data != output_data) {
+    // TODO: Run benchmarks to determine if a dedicated kernel doing data copy will
+    // be faster than invoking cudaMemcpy ?
+    CUDA_RETURN_IF_ERROR(
+        cudaMemcpyAsync(output_data, input_data, input_tensor->SizeInBytes(),
+                        cudaMemcpyDeviceToDevice, Stream(context)));
+  }
+
+  // Bail out early
+  if (indices_shape.Size() == 0) {
+    return Status::OK();
+  }
+
+  auto last_index_dimension = indices_shape[indices_shape.NumDimensions() - 1];
+
+  // We need element counts for each dimension and the input dim value for each dimension
+  // for the range [0, last_index_dimension).
+  // To avoid multiple GPU data transfers, we combine this into one array and send it through
+  TensorPitches input_strides(input_shape);
+  std::vector<int64_t> element_counts_and_input_dims(last_index_dimension * 2, 0LL);
+  for (int64_t i = 0; i < last_index_dimension; ++i) {
+    element_counts_and_input_dims[i] = input_strides[i];
+    element_counts_and_input_dims[i + last_index_dimension] = input_shape[i];
+  }
+  CudaAsyncBuffer<int64_t> element_counts_and_input_dims_gpu(this, element_counts_and_input_dims);
+  ORT_RETURN_IF_ERROR(element_counts_and_input_dims_gpu.CopyToGpu(context->GetComputeStream()));
```

Comment:
I think this should be Stream(context) exverywhere

Reply:
I tried but it does not work in this case. CopyToGpu is almost always called that way.